hbase-server/src/main/java/org/apache/hadoop/hbase/ipc/NettyRpcServer.java

```diff
@@ -181,4 +204,80 @@ public int getNumOpenConnections() {
     // allChannels also contains the server channel, so exclude that from the count.
     return channelsCount > 0 ? channelsCount - 1 : channelsCount;
   }
+
+  private synchronized void initSSL(ChannelPipeline p, boolean supportPlaintext) throws
+    X509Exception {
+    SslContext nettySslContext;
+
+    SSLContextAndOptions sslContextAndOptions = x509Util.getDefaultSSLContextAndOptions();
+    nettySslContext = sslContextAndOptions
+      .createNettyJdkSslContext(sslContextAndOptions.getSSLContext(), false);
+
+    if (supportPlaintext) {
+      p.addLast("ssl", new NettyRpcServer.DualModeSslHandler(nettySslContext));
+      LOG.debug("Dual mode SSL handler added for channel: {}", p.channel());
+    } else {
+      p.addLast("ssl", nettySslContext.newHandler(p.channel().alloc()));
+      LOG.debug("SSL handler added for channel: {}", p.channel());
+    }
+  }
+
+  /**
+   * A handler that detects whether the client would like to use
+   * TLS or not and responds in kind. The first bytes are examined
+   * for the static TLS headers to make the determination and
+   * placed back in the stream with the correct ChannelHandler
+   * instantiated.
+   */
+  class DualModeSslHandler extends OptionalSslHandler {
+
+    DualModeSslHandler(SslContext sslContext) {
+      super(sslContext);
+    }
+
+    @Override
+    protected void decode(ChannelHandlerContext context, ByteBuf in, List<Object> out)
+      throws Exception {
+      if (in.readableBytes() >= 5) {
+        super.decode(context, in, out);
+      } else if (in.readableBytes() > 0) {
+        // It requires 5 bytes to detect a proper ssl connection. In the
```

Comment:
If 5 bytes are required why `>= 5` in the previous if?

Reply:
That's copied from the base class.